### PR TITLE
Openverse: prevent multiple insertions during upload

### PR DESCRIPTION
**Pull request by swissspidy**
## What?

When adding an Openverse image from the inserter, `mediaUpload()` is called to upload the image to the media library.

Inserting the block onto the canvas happens in the `onFileChange` callback.

This callback can theoretically be called multiple times, especially in the future.

The block should only be inserted once though, and on subsequent invocations it should just update the existing block.

## Why?

I noticed this while working on client-side media processing, see https://github.com/swissspidy/media-experiments/issues/693 and #61447

There, the `onFileChange` callback can be fired dozens of times, which caused the block to be inserted over and over again.

## How?

Uses a `ref` to check whether the block has already been inserted or not.

## Testing Instructions

1. Activate the Media Experiments plugin
2. Insert an Openverse image through the media inserter
3. Verify that it's only added to the canvas once and not multiple times

### Testing Instructions for Keyboard

N/A

### CODE REVIEW

File: packages/block-editor/src/components/inserter/media-tab/media-preview.js [hunk 1]
```diff
@@ -180,18 +182,34 @@ export function MediaPreview( { media, onClick, category } ) {
...
+									},
+								} );
+
+								// TODO: Move to a new onSuccess callback as part of #61447.
+								createSuccessNotice(
+									__( 'Image uploaded and inserted.' ),
+									{ type: 'snackbar', id: 'inserter-notice' }
+								);
+
+								hasInsertedBlock.current = true;
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/media-tab/media-preview.js]: When should we reset this ref? Shouldn't we reset it to "false" when the media prop changes?

Comment (swissspidy) [reply packages/block-editor/src/components/inserter/media-tab/media-preview.js]: Ah good point, I forgot about that 😄

**Comment by youknowriad** [comment]
> The block should only be inserted once though, and on subsequent invocations it should just update the existing block.

Actually I'm not sure this is true. Isn't that different from the behavior of the inserter for blocks and patterns. Personally I expect a new block to be inserted each time I click on the media. Maybe we can save a cache of media elements per external image if needed but I'm not sure we should be changing how the inserter works.

**Comment by swissspidy** [comment]
> Actually I'm not sure this is true. Isn't that different from the behavior of the inserter for blocks and patterns. Personally I expect a new block to be inserted each time I click on the media. Maybe we can save a cache of media elements per external image if needed but I'm not sure we should be changing how the inserter works.

No, this is not what this PR is about... Let me try to explain differently.

When using Media Experiments, the issue is this:

1. You click on the media **once**
2. The image file is `fetch`ed within `onMediaInsert` and then uploaded
3. In `onMediaInsert`, a `onFileChange` callback is added to insert a block.
4. The file is compressed, resized, etc. Every such step calls `onFileChange`
5. **Because `onFileChange` is called numerous times, numerous blocks are inserted with the same image.**

**_Of course_**, every time I click on the media, a new upload process should be triggered and a new block should be added. But then this block should only be added **once** per click, not numerous times.

Hope that makes sense. If not, just install Media Experiments and try it yourself. Happy to provide a screencast otherwise...

**Comment by youknowriad** [comment]
Yes, it does, thanks for explaining. Let me test this a bit.

**Comment by youknowriad** [comment]
In my tests in this PR, If I click again on an image after I had added it to the post, it doesn't insert it again. Am I missing something?

**Comment by youknowriad** [comment]
Actually forgot about my last comment :). It's working properly now :P

**Comment by swissspidy** [comment]
> In my tests in this PR, If I click again on an image after I had added it to the post, it doesn't insert it again. Am I missing something?

Well then this unintentional, and probably is because of https://github.com/WordPress/gutenberg/pull/65719#discussion_r1780599379 where I didn't reset the ref.

File: packages/block-editor/src/components/inserter/media-tab/media-preview.js (hunk 1 above)
Comment (swissspidy) [reply packages/block-editor/src/components/inserter/media-tab/media-preview.js]: Removed the ref now.

**Comment by swissspidy** [comment]
I just removed the ref now in favor of simply using `getBlock`. Now the insertion works as expected.

File: packages/block-editor/src/components/inserter/media-tab/media-preview.js [hunk 2]
```diff
@@ -180,18 +181,32 @@ export function MediaPreview( { media, onClick, category } ) {
...
+								onClick( {
+									...clonedBlock,
+									attributes: {
+										...clonedBlock.attributes,
+										id: img.id,
+										url: img.url,
+									},
+								} );
+
+								// TODO: Move to a new onSuccess callback as part of #61447.
```
Comment (youknowriad) [inline packages/block-editor/src/components/inserter/media-tab/media-preview.js]: Not sure the comment is needed for now.

**Review by youknowriad — APPROVED** [review]
I guess it's very hard to trigger the "updateBlockAttributes" path at the moment. But this looks good to me and tests well.

**Comment by swissspidy** [comment]
> I guess it's very hard to trigger the "updateBlockAttributes" path at the moment. But this looks good to me and tests well.

Yeah right now it's only with the Media Experiments plugin, but eventually also with the client-side media processing experiment flag in Gutenberg (once that is merged)

**Pull request merged**